### PR TITLE
Clear keyboard and mouse states on detach and focus loss

### LINKED ISSUE #4005: Keyboard's detach method does not clear pressed keys
Keyboard's detach method should clear the currently pressed keys. Otherwise if someone was holding down any key while the call were made, it would lead to keyboard.isPressed(key) to return true even after detachment. The user may then un-press a key with the keyboard detached, not triggering an event, and finally after the keyboard were to be attached again it would think that the user is continuously pressing the key again.

Add this line at the very end of the detach method body to paliate this issue.

```js 
this._keymap = {};
```

**Note**: Perhaps you instead want to porpusedfully trigger a **"key up"** event for every key currently being pressed whenever you detach the keyboard, so that scripts that have subscribed to that event can be notified. Either way, **the key map should be empty after detachment** (if there is no keyboard, how can any key be pressed?).

keyboard::detach() method:
https://github.com/playcanvas/engine/blob/d89c6d5dbb0719d5c17963cd7a37c7aeb137dddf/src/input/keyboard.js#L151-L167

### PR DESCRIPTION
**Pull request by mvaligursky**
Fixes #4005. Keys and mouse buttons can remain stuck when released while detached; mouse buttons can also remain stuck when released after the window loses focus.

**Changes:**
- Clear current and previous input states when detaching the keyboard or mouse.
- Clear mouse button states on window blur and when the document becomes hidden, matching keyboard behavior.
- Preserve the existing behavior of emitting no synthetic keyup or mouseup events.
- Document attachment behavior and cover detachment, focus loss, listener cleanup, reattachment and fresh presses with regression tests.

**API Changes:**
`Keyboard.detach()` and `Mouse.detach()` now reset `isPressed()`, `wasPressed()`, and `wasReleased()` to false. Mouse state also resets on window blur or document hiding.

`Keyboard.attach()` resets key states when already attached, including when passed the same element. `Mouse.attach()` continues to preserve button states when changing targets.

Before → after, assuming the input is held:
- `keyboard.detach(); keyboard.isPressed(KEY_UP);` — `true` → `false`.
- `mouse.detach(); mouse.isPressed(MOUSEBUTTON_LEFT);` — `true` → `false`.
- `keyboard.attach(otherElement); keyboard.isPressed(KEY_UP);` — `true` → `false`.
- `mouse.isPressed(MOUSEBUTTON_LEFT)` after window blur or document hiding — `true` → `false`.

### CODE REVIEW

File: src/platform/input/keyboard.js [hunk 1]
```diff
@@ -232,6 +233,9 @@ class Keyboard extends EventHandler {
 
         document.removeEventListener('visibilitychange', this._visibilityChangeHandler, false);
         window.removeEventListener('blur', this._windowBlurHandler, false);
+
+        this._keymap = {};
```
Comment (mvaligursky) [inline src/platform/input/keyboard.js]: These two lines are byte-identical to the whole body of `_handleWindowBlur()` (line ~352), which is not a coincidence — detaching and losing focus need exactly the same treatment, for exactly the same reason.

Worth collapsing to one site (`this._handleWindowBlur()` here, or both calling a shared `_clearState()`), because the duplication is the kind that decays asymmetrically: the first time this clearing needs to grow — a modifier cache, a repeat-state flag, an event — one of the two copies gets it. Reusing it also states the intent in code, that a detach is treated as a blur, which is the same reasoning that justifies not firing synthetic `keyup`s.

File: src/platform/input/mouse.js [hunk 2]
```diff
@@ -181,6 +182,9 @@ class Mouse extends EventHandler {
         window.removeEventListener('mousedown', this._downHandler, options);
         window.removeEventListener('mousemove', this._moveHandler, options);
         window.removeEventListener('wheel', this._wheelHandler, options);
+
+        this._buttons.fill(false);
```
Comment (mvaligursky) [inline src/platform/input/mouse.js]: Adjacent gap, worth knowing about even if it's out of scope: `Mouse` has no blur or visibilitychange handling at all — I grepped, there's nothing — so the stuck-button bug this PR fixes for detach still happens on focus loss.

`mouseup` is bound on `window`, so a release anywhere in the page is caught, but a release that happens after the window loses focus (alt-tab, a native dialog, dragging out of the browser) never arrives, and nothing clears `_buttons`. `Keyboard` grew `_handleWindowBlur`/`_handleVisibilityChange` for precisely that reason; `Mouse` never did.

Now that the clearing exists here, closing it is close to free — the same `fill(false)` pair behind a blur/visibilitychange listener registered in `attach` and removed in `detach`. Reasonable to leave for a follow-up, but it's the other half of #4005's family and the person best placed to notice it is whoever is already in this file.

File: src/platform/input/keyboard.js [hunk 3]
```diff
@@ -232,6 +233,9 @@ class Keyboard extends EventHandler {
```
Comment (mvaligursky) [inline src/platform/input/keyboard.js:235]: The API Changes section lists `detach()`, but `attach()` also clears held keys now: it calls `this.detach()` when re-targeting (line ~208), so `keyboard.attach(otherElement)` mid-hold drops the key state. The new test pins this, so it's clearly intended — it just isn't in the API notes, and "switching the element I listen on resets my held keys" is the more surprising of the two for a caller who only read that `detach()` changed.

It also splits the two APIs this PR touches: `Mouse.attach()` never calls `detach()`, so re-attaching a mouse keeps its held buttons while re-attaching a keyboard doesn't. Pre-existing structural difference rather than anything introduced here, but it becomes observable for the first time with this change, so it's worth either noting or aligning.

**Review by mvaligursky — COMMENTED** [review]
🤖 **Automated PR review** — posted on my behalf by Claude Code (Opus 4.8). Not a human review. The points below are suggestions to weigh as possible improvements, not changes that necessarily need to be addressed.

---

The fix is right and the "no synthetic keyup" choice has precedent: `_handleWindowBlur` already clears both maps without firing anything, so detach now behaves the same way as losing focus, which is the consistent call. 27 input tests pass, lint clean.

Three points inline — one duplication, one adjacent gap, one undocumented consequence.

File: src/platform/input/mouse.js [hunk 4]
```diff
@@ -319,6 +343,27 @@ class Mouse extends EventHandler {
...
+            this._handleWindowBlur();
+        }
+    }
+
+    /**
+     * Handle the browser blur event.
+     *
+     * @private
+     */
+    _handleWindowBlur() {
```
Comment (mvaligursky) [inline src/platform/input/mouse.js:362]: Worth knowing about now that this hook exists: buttons were not the only mouse state that survives focus loss wrongly. `_lastX`/`_lastY` do too, and they feed the movement delta — `MouseEvent` computes `dx = this.x - mouse._lastX` on the non-pointer-locked path (mouse-event.js ~153), with `_handleMove` updating them unguarded at the end of every move.

So: pointer sits at (100, 100), window loses focus, the user moves the physical mouse across the screen or to another monitor, focus returns, and the first `mousemove` reports (900, 700) → `dx = 800, dy = 600`. An orbit or fly camera consuming `event.dx/dy` — the standard pattern — snaps hard on the first frame back. Same shape as the bug this PR fixes: stale input state that outlives the focus it was captured under.

Two caveats if you pick it up, and the first is why I'd raise it here rather than let someone "fix" it later by pattern-matching:

- **Zeroing `_lastX`/`_lastY` in this method would make it worse, not better** — the next delta becomes `x - 0`, i.e. the full coordinate. It needs a validity flag instead: clear it on blur, and on the first `_handleMove` afterwards seed `_lastX`/`_lastY` from the event and report `dx = dy = 0` before marking it valid again.
- Only the **unlocked** path is affected; under pointer lock `dx`/`dy` come from `event.movementX`, which the browser computes per event and can't go stale.

Pre-existing and reasonable to leave out of this PR — but it's the other half of "input state doesn't survive focus loss", and after this commit `_handleWindowBlur` is exactly where it belongs.

**Review by mvaligursky — COMMENTED** [review]
🤖 **Automated re-review** (commit `b9e5e62`) — posted on my behalf by Claude Code (Opus 4.8). Not a human review. The points below are suggestions to weigh as possible improvements, not changes that necessarily need to be addressed.

---

All three points resolved, and the mouse focus handling mirrors the keyboard cleanly — listeners registered in `attach`, removed in `detach`, `_handleWindowBlur` as the single clearing site for both classes. I checked the one hazard the new registration could have introduced: `Mouse.attach` has `if (this._attached) return;` after updating `_target`, so re-targeting can't double-register the blur/visibilitychange listeners. 33 tests pass (up from 27), lint clean.

Documenting the `attach()` asymmetry rather than aligning it is a fair call, and the mutual `{@link}` cross-references make the difference discoverable from either side, which was the part that mattered.

One adjacent item, inline — same failure family, and the hook for it now exists.
